{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "conda environment tf2024"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prosit dataset\n",
    "\n",
    "https://figshare.com/articles/dataset/ProteomeTools_non_tryptic_-_Prosit_fragmentation_-_Data/12937092"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['collision_energy', 'collision_energy_aligned_normed', 'collision_energy_normed', 'intensities_pred', 'intensities_raw', 'masses_raw', 'method', 'precursor_charge_onehot', 'rawfile', 'reverse', 'scan_number', 'score', 'sequence_integer', 'spectral_angle']\n",
      "9937728\n",
      "9937728\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(9937728, 174)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import h5py\n",
    "\n",
    "d = h5py.File(\"/d0/tp/data_transformer/prosit_download/prediction_hcd_train.hdf5\", \"r\")\n",
    "\n",
    "ls = list(d.keys())\n",
    "print(ls)\n",
    "\n",
    "print(len(d['collision_energy']))\n",
    "\n",
    "print(len(d['intensities_pred']))\n",
    "\n",
    "x = np.array(d['sequence_integer'][:])\n",
    "x.shape\n",
    "\n",
    "y = np.array(d['intensities_raw'][:])\n",
    "y.shape\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## visualize y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code from (or adapted from) https://github.com/kusterlab/prosit/ Apache License 2.0\n",
    "# See README.md\n",
    "\n",
    "import numpy as np\n",
    "import pyteomics\n",
    "from pyteomics import mass  # pylint: disable=unused-import\n",
    "\n",
    "from aiproteomics.e2e import constants\n",
    "from aiproteomics.e2e import utils\n",
    "\n",
    "\n",
    "def generate_aa_comp():\n",
    "    \"\"\"\n",
    "    >>> aa_comp = generate_aa_comp()\n",
    "    >>> aa_comp[\"M\"]\n",
    "    Composition({'H': 9, 'C': 5, 'S': 1, 'O': 1, 'N': 1})\n",
    "    >>> aa_comp[\"Z\"]\n",
    "    Composition({'H': 9, 'C': 5, 'S': 1, 'O': 2, 'N': 1})\n",
    "    \"\"\"\n",
    "\n",
    "    db = pyteomics.mass.mass.Unimod()\n",
    "    aa_comp = dict(pyteomics.mass.std_aa_comp)\n",
    "    s = db.by_title(\"Oxidation\")[\"composition\"]\n",
    "    aa_comp[\"Z\"] = aa_comp[\"M\"] + s\n",
    "    s = db.by_title(\"Carbamidomethyl\")[\"composition\"]\n",
    "    aa_comp[\"C\"] = aa_comp[\"C\"] + s\n",
    "    return aa_comp\n",
    "\n",
    "\n",
    "aa_comp = generate_aa_comp()\n",
    "\n",
    "\n",
    "def get_ions():\n",
    "    x = np.empty(\n",
    "        [constants.MAX_ION, len(constants.ION_TYPES), constants.MAX_FRAG_CHARGE],\n",
    "        dtype=\"|S6\",\n",
    "    )\n",
    "    for fz in range(constants.MAX_FRAG_CHARGE):\n",
    "        for fty_i, fty in enumerate(constants.ION_TYPES):\n",
    "            for fi in range(constants.MAX_ION):\n",
    "                ion = fty + str(fi + 1)\n",
    "                if fz > 0:\n",
    "                    ion += \"({}+)\".format(fz + 1)\n",
    "                x[fi, fty_i, fz] = ion\n",
    "    x.flatten()\n",
    "    return x\n",
    "\n",
    "\n",
    "ox_int = constants.ALPHABET[\"M(ox)\"]\n",
    "c_int = constants.ALPHABET[\"C\"]\n",
    "\n",
    "\n",
    "def calculate_mods(sequence_integer):\n",
    "    \"\"\"\n",
    "    >>> x = np.array([2, 15, 4, 3, 0, 0])\n",
    "    >>> calculate_mods(x)\n",
    "    1\n",
    "    >>> x = np.array([2, 15, 21, 3, 0, 0])\n",
    "    >>> calculate_mods(x)\n",
    "    2\n",
    "    \"\"\"\n",
    "    return len(np.where((sequence_integer == ox_int) | (sequence_integer == c_int))[0])\n",
    "\n",
    "\n",
    "def generate_mods_string_tuples(sequence_integer):\n",
    "    list_mods = []\n",
    "    for mod in [ox_int, c_int]:\n",
    "        for position in np.where(sequence_integer == mod)[0]:\n",
    "            if mod == c_int:\n",
    "                list_mods.append((position, \"C\", \"Carbamidomethyl\"))\n",
    "            elif mod == ox_int:\n",
    "                list_mods.append((position, \"M\", \"Oxidation\"))\n",
    "            else:\n",
    "                raise ValueError(\"cant be true\")\n",
    "    list_mods.sort(key=lambda tup: tup[0])  # inplace\n",
    "    return list_mods\n",
    "\n",
    "\n",
    "def generate_mod_strings(sequence_integer):\n",
    "    \"\"\"\n",
    "    >>> x = np.array([1,2,3,1,2,21,0])\n",
    "    >>> y, z = generate_mod_strings(x)\n",
    "    >>> y\n",
    "    '3/1,C,Carbamidomethyl/4,C,Carbamidomethyl/5,M,Oxidation'\n",
    "    >>> z\n",
    "    'Carbamidomethyl@C2; Carbamidomethyl@C5; Oxidation@M6'\n",
    "    \"\"\"\n",
    "    list_mods = generate_mods_string_tuples(sequence_integer)\n",
    "    if len(list_mods) == 0:\n",
    "        return \"0\", \"\"\n",
    "\n",
    "    returnString_mods = \"\"\n",
    "    returnString_modString = \"\"\n",
    "    returnString_mods += str(len(list_mods))\n",
    "    for i, mod_tuple in enumerate(list_mods):\n",
    "        returnString_mods += \"/\" + str(mod_tuple[0]) + \",\" + mod_tuple[1] + \",\" + mod_tuple[2]\n",
    "        if i == 0:\n",
    "            returnString_modString += mod_tuple[2] + \"@\" + mod_tuple[1] + str(mod_tuple[0] + 1)\n",
    "        else:\n",
    "            returnString_modString += (\n",
    "                \"; \" + mod_tuple[2] + \"@\" + mod_tuple[1] + str(mod_tuple[0] + 1)\n",
    "            )\n",
    "\n",
    "    return returnString_mods, returnString_modString\n",
    "\n",
    "\n",
    "def convert_to_speclib(data, fmt='tsv', intensity_scaling=10000, unknown_value_str='NA'):\n",
    "    \"\"\"\n",
    "    Convert the given data frame of prediction data into a speclib format.\n",
    "    Supported formats are currently 'msp' and 'tsv'\n",
    "    \"\"\"\n",
    "\n",
    "    IONS = get_ions().reshape(174, -1).flatten()\n",
    "\n",
    "    speclibtxt = ''\n",
    "    for i in range(data[\"iRT\"].shape[0]):\n",
    "        # Get the intensity, ion and mass for all positive intensities\n",
    "        aIntensity = data[\"intensities_pred\"][i]\n",
    "        sel = np.where(aIntensity > 0)\n",
    "        aIntensity = aIntensity[sel]\n",
    "        aIons = IONS[sel]\n",
    "        aMass = data[\"masses_pred\"][i][sel]\n",
    "\n",
    "        # Get the original inputs to the fragmentation model\n",
    "        collision_energy = data[\"collision_energy_aligned_normed\"][i] * 100\n",
    "        precursor_charge = data[\"precursor_charge_onehot\"][i].argmax() + 1\n",
    "        sequence_integer = data[\"sequence_integer\"][i]\n",
    "\n",
    "        # Get the predicted normalized retention time and ion mobility\n",
    "        # for this sequence (if available)\n",
    "        if \"iRT\" in data:\n",
    "            iRT = np.squeeze(data[\"iRT\"][i])\n",
    "        else:\n",
    "            iRT = None\n",
    "\n",
    "        if \"ccs\" in data:\n",
    "            ccs = np.squeeze(data[\"ccs\"][i])\n",
    "        else:\n",
    "            ccs = None\n",
    "\n",
    "        # If available, use the protein id and gene name in the spec lib\n",
    "        if \"protein_id\" in data:\n",
    "            protein_id = data[\"protein_id\"][i]\n",
    "        else:\n",
    "            protein_id = unknown_value_str\n",
    "\n",
    "        if \"gene_name\" in data:\n",
    "            gene_name = data[\"gene_name\"][i]\n",
    "        else:\n",
    "            gene_name = unknown_value_str\n",
    "\n",
    "        # Build a predicted Spectrum representation for this sequence\n",
    "        spec = Spectrum(\n",
    "            aIntensity,\n",
    "            collision_energy,\n",
    "            iRT,\n",
    "            ccs,\n",
    "            aMass,\n",
    "            precursor_charge,\n",
    "            sequence_integer,\n",
    "            aIons,\n",
    "            protein_id,\n",
    "            gene_name,\n",
    "            intensity_scaling,\n",
    "            unknown_value_str\n",
    "        )\n",
    "\n",
    "        # Output string representation of the spectrum in the chosen format\n",
    "        if fmt == 'msp':\n",
    "            speclibtxt += spec.to_msp()\n",
    "        elif fmt == 'tsv':\n",
    "            speclibtxt += spec.to_tsv()\n",
    "        else:\n",
    "            raise ValueError('Format not supported. Supported formats are msp and tsv')\n",
    "\n",
    "    return speclibtxt\n",
    "\n",
    "\n",
    "class Spectrum:\n",
    "    def __init__(\n",
    "        self,\n",
    "        aIntensity,\n",
    "        collision_energy,\n",
    "        iRT,\n",
    "        ccs,\n",
    "        aMass,\n",
    "        precursor_charge,\n",
    "        sequence_integer,\n",
    "        aIons,\n",
    "        protein_id,\n",
    "        gene_name,\n",
    "        intensity_scaling,\n",
    "        unknown_value_str='NA'\n",
    "    ):\n",
    "        self.intensity_scaling = intensity_scaling\n",
    "\n",
    "        self.aIntensity = aIntensity\n",
    "        self.collision_energy = collision_energy\n",
    "\n",
    "        if iRT:\n",
    "            self.iRT = str(iRT)\n",
    "        else:\n",
    "            self.iRT = unknown_value_str\n",
    "\n",
    "        if ccs:\n",
    "            self.ccs = str(ccs)\n",
    "        else:\n",
    "            self.ccs = unknown_value_str\n",
    "\n",
    "        self.aMass = aMass\n",
    "        self.precursor_charge = precursor_charge\n",
    "        self.aIons = aIons\n",
    "        self.mod, self.mod_string = generate_mod_strings(sequence_integer)\n",
    "        self.sequence = utils.get_sequence(sequence_integer)\n",
    "        # amino acid Z which is defined at the toplevel in generate_aa_comp\n",
    "        self.precursor_mass = pyteomics.mass.calculate_mass(\n",
    "            self.sequence.replace(\"M(ox)\", \"Z\"),\n",
    "            aa_comp=aa_comp,\n",
    "            ion_type=\"M\",\n",
    "            charge=int(self.precursor_charge),\n",
    "        )\n",
    "        self.protein_id = protein_id\n",
    "        self.gene_name = gene_name\n",
    "\n",
    "\n",
    "    def to_msp(self):\n",
    "        s = \"Name: {sequence}/{charge}\\nMW: {precursor_mass}\\n\"\n",
    "        s += \"Comment: Parent={precursor_mass} Collision_energy={collision_energy} \"\n",
    "        s += \"Mods={mod} ModString={sequence}//{mod_string}/{charge} \"\n",
    "        s += \"iRT={iRT}\"\n",
    "        s += \"ccs={ccs}\"\n",
    "        s += \"\\nNum peaks: {num_peaks}\"\n",
    "        num_peaks = len(self.aIntensity)\n",
    "        s = s.format(\n",
    "            sequence=self.sequence.replace(\"M(ox)\", \"M\"),\n",
    "            charge=self.precursor_charge,\n",
    "            precursor_mass=self.precursor_mass,\n",
    "            collision_energy=np.round(self.collision_energy[0], 0),\n",
    "            mod=self.mod,\n",
    "            mod_string=self.mod_string,\n",
    "            iRT=self.iRT,\n",
    "            num_peaks=num_peaks,\n",
    "        )\n",
    "        for mz, intensity, ion in zip(self.aMass, self.aIntensity, self.aIons):\n",
    "            s += \"\\n\" + str(mz) + \"\\t\" + str(intensity) + '\\t\"'\n",
    "            s += ion.decode(\"UTF-8\").replace(\"(\", \"^\").replace(\"+\", \"\") + '/0.0ppm\"'\n",
    "        return s + '\\n'\n",
    "\n",
    "\n",
    "    def to_tsv(self):\n",
    "        # Remove oxidation notation from sequence string\n",
    "        peptide_seq = self.sequence.replace(\"M(ox)\", \"M\")\n",
    "\n",
    "        s = ''\n",
    "\n",
    "        # Add rows for each fragment\n",
    "        for mz, intensity, ion in zip(self.aMass, self.aIntensity, self.aIons):\n",
    "            # TODO: Make this values->string->values more sane\n",
    "            ion_str = ion.decode('UTF-8')\n",
    "            fragment_type = ion_str[0]\n",
    "            fragment_series_number = ion_str.split('(')[0][1:]\n",
    "            if '(' in ion_str:\n",
    "                fragment_charge = ion_str.split('(')[1].split('+')[0] # This is horrible\n",
    "            else:\n",
    "                fragment_charge = '1'\n",
    "            annotation = fragment_type + fragment_series_number + '^' + fragment_charge\n",
    "\n",
    "            # Scale intensity (often the highest peak is 10000 by convention)\n",
    "            intensity *= self.intensity_scaling\n",
    "\n",
    "            # Collect the values for each column describing a particular peak\n",
    "            row = [\n",
    "                str(self.precursor_mass), \n",
    "                str(mz),\n",
    "                annotation,\n",
    "                str(self.protein_id),\n",
    "                str(self.gene_name),\n",
    "                peptide_seq,\n",
    "                self.sequence,\n",
    "                str(self.precursor_charge),\n",
    "                str(intensity),\n",
    "                self.iRT,\n",
    "                self.ccs,\n",
    "                fragment_type,\n",
    "                fragment_charge,\n",
    "                fragment_series_number,\n",
    "                'NA'\n",
    "            ]\n",
    "            # Add the row to the output string\n",
    "            s += '\\t'.join(row) + '\\n'\n",
    "            \n",
    "        return s\n",
    "\n",
    "def get_tsv_format_header():\n",
    "    tsv_columns = [\n",
    "        \"PrecursorMz\",\n",
    "        \"ProductMz\",\n",
    "        \"Annotation\",\n",
    "        \"ProteinId\",\n",
    "        \"GeneName\",\n",
    "        \"PeptideSequence\",\n",
    "        \"ModifiedPeptideSequence\",\n",
    "        \"PrecursorCharge\",\n",
    "        \"LibraryIntensity\",\n",
    "        \"NormalizedRetentionTime\",\n",
    "        \"PrecursorIonMobility\",\n",
    "        \"FragmentType\",\n",
    "        \"FragmentCharge\",\n",
    "        \"FragmentSeriesNumber\",\n",
    "        \"FragmentLossType\"\n",
    "    ]\n",
    "    return '\\t'.join(tsv_columns) + '\\n'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c0 = np.array(d['precursor_charge_onehot'])[0]\n",
    "\n",
    "x0 = x[0]\n",
    "\n",
    "y0 = y[0]\n",
    "\n",
    "y0.shape\n",
    "\n",
    "int2seq = '_ACDEFGHIKLMNPQRSTVWYo'        \n",
    "int2seqf = lambda x : ''.join([int2seq[c] for c in x if c > 0])\n",
    "int2seqf(x0)\n",
    "\n",
    "\n",
    "# prepare\n",
    "\n",
    "from aiproteomics.e2e import sanitize\n",
    "from aiproteomics.e2e import tensorize\n",
    "import pandas as pd\n",
    "\n",
    "peptide_data = {\"sequence_integer\" : np.reshape(x0, (1, -1)),\n",
    "                \"precursor_charge_onehot\": np.reshape(c0, (1, -1))}                \n",
    "            \n",
    "prediction = y0\n",
    "peptide_data[\"intensities_pred\"] = prediction\n",
    "#peptide_data = sanitize.sanitize_prediction_output(peptide_data)\n",
    "intensities = np.reshape(y0, (1, -1))\n",
    "intensities[intensities < 0] = 0\n",
    "intensities = sanitize.normalize_base_peak(intensities)\n",
    "\n",
    "intensities = sanitize.reshape_dims(intensities)\n",
    "print(intensities.shape)\n",
    "\n",
    "sequence_lengths = np.count_nonzero(peptide_data[\"sequence_integer\"], axis=1)\n",
    "intensities = sanitize.mask_outofrange(intensities, sequence_lengths)\n",
    "\n",
    "charges = list(peptide_data[\"precursor_charge_onehot\"].argmax(axis=1) + 1)\n",
    "intensities = sanitize.mask_outofcharge(intensities, charges)\n",
    "\n",
    "# Shape = N_sequences x 174\n",
    "intensities = sanitize.reshape_flat(intensities)\n",
    "peptide_data[\"intensities_pred\"] = intensities\n",
    "\n",
    "peptide_data[\"iRT\"] = np.array([0])\n",
    "peptide_data[\"ccs\"] = np.array([0])\n",
    "peptide_data[\"collision_energy_aligned_normed\"] = np.array([0])\n",
    "\n",
    "df = pd.DataFrame.from_dict({\"modified_sequence\" : [int2seqf(x0)], \"precursor_charge\" : charges})\n",
    "\n",
    "masses_pred = tensorize.get_mz_applied(df)\n",
    "\n",
    "from aiproteomics.e2e.constants import MAX_FRAG_CHARGE, MAX_NLOSSES\n",
    "masses_pred = sanitize.cap(masses_pred, MAX_NLOSSES, MAX_FRAG_CHARGE)\n",
    "\n",
    "# Calculate length of each (integer) peptide sequence\n",
    "lengths = (peptide_data[\"sequence_integer\"] > 0).sum(1)\n",
    "masses_pred = sanitize.mask_outofrange(masses_pred, lengths)\n",
    "masses_pred = sanitize.mask_outofcharge(masses_pred, df.precursor_charge)\n",
    "\n",
    "\n",
    "# Output shape: N_sequences x 174\n",
    "masses_pred = sanitize.reshape_flat(masses_pred)\n",
    "peptide_data[\"masses_pred\"] = masses_pred\n",
    "\n",
    "#convert_to_speclib\n",
    "\n",
    "from aiproteomics.e2e import speclib_writers\n",
    "speclibtxt = speclib_writers.convert_to_speclib(peptide_data, fmt='tsv')\n",
    "\n",
    "#plot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# end here\n",
    "\n",
    "aMass = data[\"masses_pred\"][i][sel]\n",
    "IONS = get_ions().reshape(174, -1).flatten()\n",
    "\n",
    "# Get the intensity, ion and mass for all positive intensities\n",
    "aIntensity = y0\n",
    "sel = np.where(aIntensity > 0)\n",
    "aIntensity = aIntensity[sel]\n",
    "aIons = IONS[sel]\n",
    "aMass = peptide_data[\"masses_pred\"][0][sel]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MS/MS plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(peptide_data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.14 ('tf2024')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "37e3d7ca35469379dac3af466c5e2d3a10b0ab855cbd59bd4a92a783d05bffb6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
